### PR TITLE
Added Docker support for Talawa Admin repository

### LINKED ISSUE #1075: Talawa-Admin Docker Support Issue
### Talawa-Admin Docker Support Issue:

**Title:**
Adding Docker Support to Talawa-Admin

**Description:**

**Problem:**
Talawa-Admin currently lacks Docker support, hindering its ease of deployment and scalability. Adding Docker support is essential to streamline the deployment process and make it more accessible to a broader audience.

**Solution:**
Implementing Docker support for Talawa-Admin would involve creating a Dockerfile and configuration to package the application into a Docker container. This containerization would simplify deployment, improve consistency across environments, and enhance scalability.

**Approach:**
To achieve Docker support for Talawa-Admin, we can follow these steps:
1. Create a Dockerfile specifying the necessary dependencies and configuration.
2. Configure the application to work seamlessly within a Docker container.
3. Document the steps for building and running Talawa-Admin using Docker in the project's README.

**Additional Context:**
Adding Docker support aligns with modern development practices, making it easier for developers to set up and run Talawa-Admin in various environments. Docker containers provide isolation, portability, and reproducibility, contributing to a more robust and user-friendly development and deployment experience.

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Devesh326**
**What kind of change does this PR introduce?**

Docker feature

**Issue Number:**

Fixes #1075

**Did you add tests for your changes?**

No

**If relevant, did you update the documentation?**

Will update the documentation once the solution is approved.

**Summary**

Dockerization of a repository has the following advantages:
1. Reduces Setup Issues for Contributors
2. Eliminates "It Works on My Machine" Problem
3. Ensures Consistency Across Environments

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Update this file with the newly added docker files
    1. https://github.com/PalisadoesFoundation/talawa-admin/blob/develop/.github/workflows/authorized-changes-detection.yml
1. Update `INSTALLATION.md`. Users will need to know what to expect. Use the version in `Talawa-API` as a guide.
2. Update `setup.ts`. Use the version in `Talawa-API` as a guide. This feature must be easy to implement.

**Comment by palisadoes** [comment]
1. @kb-0311
1. @vasujain275
1. @EshaanAgg 

Please take a look and comment. You have been involved with Docker updates to the repos in the recent past

**Comment by Devesh326** [comment]
Hey @palisadoes,
There's already an PR for script.ts #1350 , so should i work on it once its merged to avoid merge conflicts?

**Comment by vasujain275** [comment]
@Devesh326 @palisadoes

Wouldn't it be better to push the talawa-admin image to Docker Hub or any public Docker repository and then use that image in the docker-compose file for talawa-api? This way, the talawa-admin container will be in the same Docker network/group with other containers such as mongo, redis, and the talwa-api container. We can have a CI/CD pipeline to push a new image to Docker Hub on every merge, which can be pulled by the docker-compose file in the talawa-api repo.

**Comment by Devesh326** [comment]
Hey 

> @Devesh326 @palisadoes
> 
> Wouldn't it be better to push the talawa-admin image to Docker Hub or any public Docker repository and then use that image in the docker-compose file for talawa-api? This way, the talawa-admin container will be in the same Docker network/group with other containers such as mongo, redis, and the talwa-api container. We can have a CI/CD pipeline to push a new image to Docker Hub on every merge, which can be pulled by the docker-compose file in the talawa-api repo.


While @palisadoes can provide a additional insights  but in my opinion 
1. This would increase the build/execution time for talawa-api. 
2. Definitely including docker image for talawa-admin would reduce installation/execution complexity but is there any other specific reason for both to be on the same network? 

Pls correct me if i am wrong.

**Comment by palisadoes** [comment]
1. Please fix the conflicting files
2. The setup script PR has been merged. You can add the script updates to this PR

**Comment by palisadoes** [comment]
I'm not very familiar with Docker Hub but it appears to be similar to an app store.

1. The repos are too dynamic to have such a tight coupling between Admin and API.
2. We are constantly adding and removing features in one and/or the other. 
3. This is something we can consider later when:
    1. the compatibilities between the two are more stable. 
    2. we feel confident that the `main` branches are ready for Alpha and Beta testing.

The short term aim of dockerization is to make life of the developer less complicated in getting the app setup.

**Comment by vasujain275** [comment]
@palisadoes @Devesh326

1. Yes, I agree; it is better to have separate docker-compose files for talawa-api and talawa-admin for now.
2. Once this PR is merged, we can create a different PR for setting up a github workflow to push these images to Docker Hub.
3. Once the Docker Hub CI/CD pipeline is working, we can provide developers setting up talawa-api using Docker Compose an option parameter to also run talawa-admin by using the Docker Hub image.
4. This will especially help talawa mobile app developers, as they can run both talawa-api and talawa-admin from the Docker Compose file of the talawa-api repo by passing an optional argument to the `docker-compose up` command.

**Comment by Devesh326** [comment]
Hey @palisadoes,

Something messed up when I committed the changes even though on my forked branch it says 7 files changed, here it shows 232 files changed. Should i make a new PR for this issue?

**Comment by Devesh326** [comment]
Hey @palisadoes,

If any changes are required please let me know!

**Comment by palisadoes** [comment]
> Hey @palisadoes,
> 
> When I committed the changes my forked branch says 7 files changed, but here it shows 232 files changed, maybe its showing the changed files of the setup script PR since it was mentioned here.
> 
> So should i make a new PR or this is fine?

The documentation file bug has been fixed.

1. Create a branch based on the current upstream
2. `zip` the files in the documentation directory for this branch
3. Use the PR branch
4. `unzip` the documentation over documentation directory
5. Commit the changes

Merging with your upstream should have no conflicts

**Comment by palisadoes** [comment]
@kb-0311 @vasujain275 Is this ready to merge? I'm not familiar with docker.

**Comment by vasujain275** [comment]
@Devesh326 Please fix the PR, how is this changing 232 files?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM node:20-alpine
```
Comment (vasujain275) [inline Dockerfile:1]: Use node lts instead of alpine

Comment (Devesh326) [reply Dockerfile:1]: Okay 👍

**Comment by vasujain275** [comment]
@Devesh326

1. Implement a multi-stage build for the Docker container.
2. Instead of including all the source code and Node modules in the Docker container, it should only contain build files and serve from that production build.
3. You can reference this example from here - https://www.fl0.com/blog/the-perfect-multi-stage-dockerfile-for-node-js-apps.
4. I am following the same approach and will refactor the `talawa-api` Dockerfile too once the cloud instance scripts are done, as it is more efficient and offers significant performance and security advantages.

**Comment by Devesh326** [comment]
> @Devesh326
> 
> 1. Implement a multi-stage build for the Docker container.
> 2. Instead of including all the source code and Node modules in the Docker container, it should only contain build files and serve from that production build.
> 3. You can reference this example from here - https://www.fl0.com/blog/the-perfect-multi-stage-dockerfile-for-node-js-apps.
> 4. I am following the same approach and will refactor the `talawa-api` Dockerfile too once the cloud instance scripts are done, as it is more efficient and offers significant performance and security advantages.

Sure Great Idea!

**Comment by palisadoes** [comment]
> > Hey @palisadoes,
> > 
> > When I committed the changes my forked branch says 7 files changed, but here it shows 232 files changed, maybe its showing the changed files of the setup script PR since it was mentioned here.
> > 
> > So should i make a new PR or this is fine?
> 
> The documentation file bug has been fixed.
> 
> 1. Create a branch based on the current upstream
> 2. `zip` the files in the documentation directory for this branch
> 3. Use the PR branch
> 4. `unzip` the documentation over documentation directory
> 5. Commit the changes
> 
> Merging with your upstream should have no conflicts

This is how you fix the 200+ file issue.

We had a bug in one of our GitHub actions

**Comment by Devesh326** [comment]
> 
> This is how you fix the 200+ file issue.
> 
> We had a bug in one of our GitHub actions


Yeah now its resolved, i just merged the upstream branch and now its fixed!!

Noted for future reference, thanks!

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+WORKDIR /usr/src/app
+
+COPY package* .
+
+RUN npm install --legacy-peer-deps
+
+COPY . .
+
+EXPOSE 3000
```
Comment (palisadoes) [inline Dockerfile:11]: The default port has now been changed.

**Comment by palisadoes** [comment]
@kb-0311 @vasujain275 PTAL once more

**Comment by vasujain275** [comment]
@palisadoes 

This pr is not ready to merge yet, @Devesh326 is implementing the different and more efficient method for Dockerfile, once those changes are committed I will review them and let you know

**Comment by Devesh326** [comment]
I'm in the process of implementing a multi-stage build, which is a bit new concept for me. I'm currently familiarizing myself with how it works and plan to commit the changes in the next couple of days.

**Comment by palisadoes** [comment]
Closing due to inactivity

**Pull request closed without merging**